Drawterm: 2023-03-05 -> 2023-06-27, drawterm-wayland, and audio fixes

majiru opened the pull request:
###### Description of changes

This adds support for the 'linux' drawterm target which uses wayland and pipewire.
There was some upstream issues with this build for specifically nix that have now been remediated,
so we also bump the version up to include those upstream changes.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

majiru committed:
drawterm: 2023-03-05 -> 2023-06-27

majiru committed:
drawterm: add wayland+pipewire 'linux' build target support

majiru commented:
I couldn't find too much documentation on best practices for these kind of build-time target tunable's. This attaches it to an extra argument `useConfig` but if there is a more idiomatic way to do this (or if this should be another package altogether) please let me know.

luc65r commented:
Looks good to me.

I'm really glad that new features are still implemented in and around 9front. Thank you!

SuperSandro2000 commented on pkgs/tools/admin/drawterm/default.nix:
```diff
@@ -2,36 +2,50 @@
 
   installPhase = ''
     install -Dm755 -t $out/bin/ drawterm
     install -Dm644 -t $out/man/man1/ drawterm.1
```
```suggestion
    installManPage drawterm.1
```

SuperSandro2000 commented on pkgs/tools/admin/drawterm/default.nix:
```diff
@@ -2,36 +2,50 @@
-    xorg.libXt
-  ];
+  nativeBuildInputs = {
+    linux = [ pkg-config wayland-scanner ];
+    unix = [];
```
Is this depending on meta.platforms? If so we can just use that instead.

majiru replied:
Not quite. The "unix" and "linux" here refer to the upstream drawterm build targets, not meta.platforms. The names are perhaps slightly misnomers. "unix" = x11 + OSS, "linux" = wayland + pipewire. So for nixos, which still defaults to X11. it makes more sense to use the "unix" upstream target.

majiru committed:
drawterm: fix audio using padsp

drawterm uses an oss audio backend by default for CONF=unix.
The easiest way I've seen to get this working for most systems is
to wrap it with padsp.

majiru commented:
While the darwin bulds fail on ofborg due to xcode I can confirm that manually verifying the builds locally does work

SuperSandro2000 commented on pkgs/tools/admin/drawterm/default.nix:
```diff
@@ -1,37 +1,93 @@
+    (allow file-write* (subpath "/private/var/folders"))
+  '';
+
+  makeFlags = [ "CONF=${plat}" ] ++ (if plat == "osx-cocoa" then [ "CC=clang" ] else []);
```
```suggestion
  makeFlags = [ "CONF=${plat}" ]
   ++ lib.optional (plat == "osx-cocoa") "CC=clang";
```

SuperSandro2000 commented on pkgs/tools/admin/drawterm/default.nix:
```diff
@@ -1,37 +1,93 @@
 }:
-
+let
+  plat = if stdenv.targetPlatform.isMacOS then "osx-cocoa" else config;
```
```suggestion
  plat = if stdenv.targetPlatform.isDarwin then "osx-cocoa" else config;
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2381

majiru commented on pkgs/tools/admin/drawterm/default.nix:
```diff
@@ -1,37 +1,93 @@
-  # TODO: macos
-  makeFlags = [ "CONF=unix" ];
+  postPatch = ''
+    substituteInPlace gui-cocoa/Makefile --replace 'ibtool' '/usr/bin/ibtool'
```
Yup, this is doing the opposite of what we would normally use it for. This is part of the issues with using a tool that requires the full fledged XCode graphical program (and not just the command line variant). This along with the sandbox modification make it such that this build will the systems installed XCode.

This pattern was copied from vulkan-tools at the suggestion of Lily Foster in element. And while not exactly the best solution, the xcode dependency means that hydra and other caches will not be able to build it. Perhaps a comment explaining this case would be good?

SuperSandro2000 commented:
darwin fails to build on ofborg. Not sure how to mark this correctly.

```
/usr/bin/ibtool --compile drawterm.app/MainWindow.nib MainWindow.xib
xcode-select: error: no developer tools were found at '/Applications/Xcode.app', and no install could be requested (perhaps no UI is present), please install manually from 'developer.apple.com'.
make[1]: *** [Makefile:17: drawterm.app/MainWindow.nib] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-drawterm-unstable-2023-06-27.drv-0/drawterm-36debf4/gui-cocoa'
make: *** [Makefile:75: gui-cocoa/libgui.a] Error 2
error: builder for '/nix/store/2qvmm03jan6if9sac1628k97r7sc8b96-drawterm-unstable-2023-06-27.drv' failed with exit code 2;
```

majiru commented:
> darwin fails to build on ofborg. Not sure how to mark this correctly.
> 
> ```
> /usr/bin/ibtool --compile drawterm.app/MainWindow.nib MainWindow.xib
> xcode-select: error: no developer tools were found at '/Applications/Xcode.app', and no install could be requested (perhaps no UI is present), please install manually from 'developer.apple.com'.
> make[1]: *** [Makefile:17: drawterm.app/MainWindow.nib] Error 1
> make[1]: Leaving directory '/private/tmp/nix-build-drawterm-unstable-2023-06-27.drv-0/drawterm-36debf4/gui-cocoa'
> make: *** [Makefile:75: gui-cocoa/libgui.a] Error 2
> error: builder for '/nix/store/2qvmm03jan6if9sac1628k97r7sc8b96-drawterm-unstable-2023-06-27.drv' failed with exit code 2;
> ```

Yes the ofborg and hydra builders will not be able to build this due to the XCode requirement, which is commented in the code. I have the meta property set to avoid having hydra attempt to build these

majiru commented:
Due to the complexity of the macos support, and my lack of motivation to really iterate it and find a better solution. I have for now pulled it out of this patch. If someone find discussion and would like to continue the work, I have posted the macos commit as a independant gist that can be found here: [d1f2d75d20bb6144085f2225ef59aff8f6be40b8](https://gist.github.com/majiru/f68b1dd6fe5b8afaf338140fb00717f6)
